pkg/kubecost/allocation.go

```diff
@@ -42,28 +47,29 @@ const ShareNone = "__none__"
 // Allocation is a unit of resource allocation and cost for a given window
 // of time and for a given kubernetes construct with its associated set of
 // properties.
+// TODO:CLEANUP make TotalCost a function
```

Comment:
Interesting idea. Let's discuss in-person soon. There is a test now for `UnmarshalJSON` that demonstrates that unmarshaling works just fine for all those values (i.e. we don't need to implement that function ourselves) and the `MarshalJSON` doesn't feel too cumbersome to me (feel free to push me on that) so it seems like a relatively low cost to get the behavior we want. On the other hand, creating a new type (e.g. `AllocationResponse`) would mean that we'd need to convert from one to the other in API responses, which feels more cumbersome (again, feel free to push me).

For instance, we would then need `AllocationSetResponse` and `AllocationSetRangeResponse` types, too, right? And then some conversion function that turns non-Responses into Responses. And for unmarshaling, we'd need to convert back. It just feels like a step we don't need to me.

Reply:
And to address the initial concern, I have always been worried that you could incorrectly set `TotalCost` (or, more likely _forget_ to set `TotalCost` when changing some other cost, like `ExternalCost`) and end up with out-of-sync costs. If we can say that `TotalCost` is, by definition, the total of the constituent costs, then it both safeguards us on correctness and reduces the amount of data we need to hold in memory and on disk.

Reply:
As far as `TotalCost` defined as a function goes, I'm 100% with you.

I'd definitely like to talk for a few minutes about the response idea. I'm also of the opinion that its more cumbersome to have this sort of extra intermediate step to go from internal struct -> response struct (and vice versa). This may tie in to the concept of having public API specs in some form or another. A user trying to import `cost-model` in a Go project to deserialize responses (or even a user inspecting the source to copy the struct to deserialize) might get an unexpected response if they don't see the custom `MarshalJSON`. However, if we have a clearly linked API spec for allocation then I think it mitigates this concern.

As it stands, I think this is fine to go as is.

Reply:
There's no doubt that there's something to this conversation because earlier in the week I actually did something like this in building my accuracy-comparison script haha: https://gist.github.com/nikovacevic/6e08351b93459f2bc338a8dc068507b1

I felt that it was a convenient hack for what I needed, but it didn't feel like exactly what we'd want if we were to formalize the idea of a "Response" type, as separate from the underlying type. Let's keep exploring what that would look like.

Reply:
> it didn't feel like exactly what we'd want if we were to formalize the idea of a "Response" type

Totally agree. The conclusion I'm drawing to here is that what you've got in this PR is fine, for now. At some point we may want to revisit the way we define our API and response types for consumers. Then those can be targeted commonly.